### PR TITLE
Fix functions not being kept for debugger

### PR DESCRIPTION
**Pull request by augusto2112**
Explanation: At Onone, many types of functions (anything user written, compiler generated setters and getters, etc), should be kept in the final binary so they're accessible by the debugger. They were previously being optimized out.
Scope: changes which functions should be kept for the debugger or not.
Risk: Low, affects whether a function should be optimized out or not, but not the actual logic of IRGening functions.
Testing: Added a compiler test. Modified a few tests to match the current expected behavior.
Issue: rdar://126763340
Reviewer: @adrian-prantl @al45tair @eeckstein

### CODE REVIEW

**Comment by augusto2112** [comment]
@swift-ci test

**Review by adrian-prantl — APPROVED** [review]

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/8740
@swift-ci test

File: lib/SIL/IR/SILFunction.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@
 #include "swift/AST/Availability.h"
 #include "swift/AST/Expr.h"
 #include "swift/AST/GenericEnvironment.h"
+#include "swift/AST/IRGenOptions.h"
```
Comment (slavapestov) [inline lib/SIL/IR/SILFunction.cpp:19]: I don't think it makes sense for files in lib/SIL to depend on IRGen headers. Can you find another way to plumb this through?

**Comment by adrian-prantl** [comment]
@swift-ci test windows

File: lib/SIL/IR/SILFunction.cpp (hunk 1 above)
Comment (augusto2112) [reply lib/SIL/IR/SILFunction.cpp:19]: Since this already merged in main I opened https://github.com/apple/swift/pull/73727 to address this issue

Comment (augusto2112) [reply lib/SIL/IR/SILFunction.cpp]: I cherry-picked https://github.com/apple/swift/pull/73727 here as well

**Comment by augusto2112** [comment]
@swift-ci test

**Pull request merged**